monai/metrics/surface_dice.py

```diff
@@ -0,0 +1,233 @@
+        :label: nsd
+
+    with :math:`\mathcal{D}_{Y_{b,c}}` and :math:`\mathcal{D}_{\hat{Y}_{b,c}}` being two sets of nearest-neighbor
+    distances. :math:`\mathcal{D}_{Y_{b,c}}` is computed from the predicted segmentation boundary towards the reference segmentation
+    boundary and vice-versa for :math:`\mathcal{D}_{\hat{Y}_{b,c}}`. :math:`\mathcal{D}_{Y_{b,c}}^{'}` and
+    :math:`\mathcal{D}_{\hat{Y}_{b,c}}^{'}` refer to the subsets of distances that are smaller or equal to the
+    acceptable distance :math:`\tau_c`:
+
+    .. math::
+        \mathcal{D}_{Y_{b,c}}^{'} = \{ d \in \mathcal{D}_{Y_{b,c}} \, | \, d \leq \tau_c \}.
+
+
+    In the case of a class neither being present in the predicted segmentation, nor in the reference segmentation, a nan value
+    will be returned for this class. In the case of a class being present in only one of predicted segmentation or
+    reference segmentation, the class NSD will be 0.
+
+    This implementation supports 2D images. For 3D images, please refer to DeepMind's implementation
+    https://github.com/deepmind/surface-distance.
+
+    Args:
+        y_pred: Predicted segmentation, typically segmentation model output.
+            It must be a one-hot encoded, batch-first tensor [B,C,H,W].
+        y: Reference segmentation.
+            It must be a one-hot encoded, batch-first tensor [B,C,H,W].
+        class_thresholds: List of class-specific thresholds.
+            The thresholds relate to the acceptable amount of deviation in the segmentation boundary in pixels.
+            Each threshold needs to be a finite, non-negative number.
+        include_background: Whether to skip the surface dice computation on the first channel of
+            the predicted output. Defaults to ``False``.
+        distance_metric: The metric used to compute surface distances.
+            One of [``"euclidean"``, ``"chessboard"``, ``"taxicab"``].
+            Defaults to ``"euclidean"``.
+
+    Raises:
+        ValueError: If `y_pred` and/or `y` are not PyTorch tensors.
+        ValueError: If `y_pred` and/or `y` do not have four dimensions.
+        ValueError: If `y_pred` and/or `y` have different shapes.
+        ValueError: If `y_pred` and/or `y` are not one-hot encoded
+        ValueError: If the number of channels of `y_pred` and/or `y` is different from the number of class thresholds.
+        ValueError: If any class threshold is not finite.
+        ValueError: If any class threshold is negative.
+
+    Returns:
+        Pytorch Tensor of shape [B,C], containing the NSD values :math:`\operatorname {NSD}_{b,c}` for each batch index
+        :math:`b` and class :math:`c`.
+    """
+
+    if not include_background:
+        y_pred, y = ignore_background(y_pred=y_pred, y=y)
+
+    if not isinstance(y_pred, torch.Tensor) or not isinstance(y, torch.Tensor):
+        raise ValueError("y_pred and y must be PyTorch Tensor.")
+
+    if y_pred.ndimension() != 4 or y.ndimension() != 4:
```

Comment:
do you think the function could be extended to handle 3D and this value error could be removed? I don't see it's invoking any 2d only APIs in this function...

Reply:
Of course it could be extended, but the way 3D surfaces & surface distances are handled in deep mind is different from MONAI: Instead of defining surfaces at surface voxel boundaries, surface points are located on a grid that is shifted by half of the grid spacing along each axis (cf. Fig. 10 in their [paper](https://arxiv.org/abs/1809.04430)):

![grafik](https://user-images.githubusercontent.com/4249077/161634269-8a2e1161-2da0-4f05-8dcd-08bff241e6f6.png)

I think it should be discussed more broadly whether to keep the surface computation currently used in MONAI, or to switch to the deep mind computation, or to establish a switch such that the user can decide on the computation. This choice should be made consistently for all boundary-overlap metrics. Has this topic already been discussed?

Reply:
I don't think this has been discussed cc@AReinke @csudre. Agree that having a flag to use different approximations would be nice. we can have a new ticket afterwards.

For this PR for the same reason, in 2D it's also a different approximation compared with the deepmind's? because the local lengths are assumed all 1s, but the deepmind's code has the possibility of 1 and 0.5*sqrt(2)...would be great to have the paper reference in the docstring https://arxiv.org/abs/2111.05408.

Reply:
Yes, in 2D as well this implementation is different from Deep Mind's implementation. I added a clarification to the documentation. The advantage of this implementation is that it is simpler to interpret (proportion of boundary pixels that are correct) and more comparable with how other metrics in MONAI are implemented. E.g. when assessing the dice score, segmentations are also considered at the discrete pixel level with the boundaries of the segmentation being the pixel edges instead of constructing non-discrete boundaries and thus having boundary pixels contributing only partially. Also, I find it inconsistent that the non-discrete boundaries are not used for distance calculations in Deep Mind's implementation.
I would definetely welcome expanding the implementation to 3D, but since I'm not working with 3D data, others should decide whether e.g. the interpretation of the NSD as proportion of boundary pixels that are correct is desirable or consideration of anisotropic voxel spacing is required.

Reply:
Thank you, I'm merging this one soon!